staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go

```diff
@@ -31,101 +32,119 @@ import (
+			pruneObjects, err := a.findObjectsToPrune(ctx, dynamicClient, visitedUids, metav1.NamespaceNone, restMapping)
+			if err != nil {
+				return nil, fmt.Errorf("listing %v objects for prune: %w", restMapping.GroupVersionKind.String(), err)
 			}
+			allObjects = append(allObjects, pruneObjects...)
 
 		default:
-			return fmt.Errorf("unhandled scope %q", restMapping.Scope.Name())
+			return nil, fmt.Errorf("unhandled scope %q", restMapping.Scope.Name())
 		}
 	}
 
-	return nil
+	return allObjects, nil
+}
+
+func (a *ApplySet) pruneAll(ctx context.Context, dynamicClient dynamic.Interface, visitedUids sets.Set[types.UID], deleteOptions *ApplySetDeleteOptions) error {
+	allObjects, err := a.findAllObjectsToPrune(ctx, dynamicClient, visitedUids)
+	if err != nil {
+		return err
+	}
+
+	return a.deleteObjects(ctx, dynamicClient, allObjects, deleteOptions)
 }
 
-func (p *applySetPruner) prune(ctx context.Context, namespace string, mapping *meta.RESTMapping, applyset *ApplySet) error {
-	applysetLabelSelector := applyset.LabelSelectorForMembers()
+func (a *ApplySet) findObjectsToPrune(ctx context.Context, dynamicClient dynamic.Interface, visitedUids sets.Set[types.UID], namespace string, mapping *meta.RESTMapping) ([]PruneObject, error) {
+	applysetLabelSelector := a.LabelSelectorForMembers()
 
 	opt := metav1.ListOptions{
 		LabelSelector: applysetLabelSelector,
 	}
 
 	klog.V(2).Infof("listing objects for pruning; namespace=%q, resource=%v", namespace, mapping.Resource)
-	objects, err := p.dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
+	objects, err := dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	var pruneObjects []PruneObject
 	for i := range objects.Items {
 		obj := &objects.Items[i]
 
 		uid := obj.GetUID()
-		if p.visitedUids.Has(uid) {
+		if visitedUids.Has(uid) {
 			continue
 		}
 		name := obj.GetName()
-		if p.dryRunStrategy != cmdutil.DryRunClient {
-			if err := p.delete(ctx, namespace, name, mapping); err != nil {
-				return fmt.Errorf("deleting %s/%s: %w", namespace, name, err)
+		pruneObjects = append(pruneObjects, PruneObject{
+			Name:      name,
+			Namespace: namespace,
+			Mapping:   mapping,
+			Object:    obj,
+		})
+
+	}
+	return pruneObjects, nil
+}
+
+func (a *ApplySet) deleteObjects(ctx context.Context, dynamicClient dynamic.Interface, pruneObjects []PruneObject, opt *ApplySetDeleteOptions) error {
+	for i := range pruneObjects {
+		pruneObject := &pruneObjects[i]
+
+		name := pruneObject.Name
+		namespace := pruneObject.Namespace
+		mapping := pruneObject.Mapping
+
+		if opt.DryRunStrategy != cmdutil.DryRunClient {
+			if err := a.delete(ctx, dynamicClient, namespace, name, mapping, opt); err != nil {
+				return fmt.Errorf("pruning %v %s/%s: %w", pruneObject.Mapping.GroupVersionKind.String(), namespace, name, err)
```

Comment:
nit: maybe make a `pruneObject.String()` than could prevent the slash for non-namespaced objects? (e.g. `Kind=Namespace /bar` in the test)